tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,905 @@
+        mode: str,
+    ):
+        """Make a prediction on tile/wsi.
+
+        Args:
+            wsi_idx (int): index of the tile/wsi to be processed within `self`.
+            ioconfig (IOConfigSegmentor): object which defines I/O placement during
+                inference and when assembling back to full tile/wsi.
+            loader (torch.Dataloader): loader object which return batch of data
+                to be input to model.
+            save_path (str): location to save output prediction as well as possible
+                intermediat results.
+            mode (str): `tile` or `wsi` to indicate run mode.
+
+        """
+        wsi_path = self.imgs[wsi_idx]
+        mask_path = None if self.masks is None else self.masks[wsi_idx]
+        wsi_reader, mask_reader = self.get_reader(
+            wsi_path, mask_path, mode, self.auto_generate_mask
+        )
+
+        resolution = ioconfig.highest_input_resolution
+        if (
+            isinstance(wsi_reader, VirtualWSIReader)
+            and resolution["units"] != "baseline"
+        ):
+            raise ValueError("Inference on `tile` only use `units='baseline'` !")
+        wsi_proc_shape = wsi_reader.slide_dimensions(**resolution)
+
+        # * retrieve patch and tile placement
+        # this is in XY
+        (patch_inputs, patch_outputs) = self.get_coordinates(wsi_proc_shape, ioconfig)
+        if mask_reader is not None:
+            sel = self.filter_coordinates(mask_reader, patch_outputs, **resolution)
+            patch_outputs = patch_outputs[sel]
+            patch_inputs = patch_inputs[sel]
+
+        # modify the shared space so that we can update worker info without
+        # needing to re-create the worker. There should be no race-condition because
+        # only the following enumerate loop triggers the parallelism, and this portion
+        # is still in sequential execution order
+        patch_inputs = torch.from_numpy(patch_inputs).share_memory_()
+        patch_outputs = torch.from_numpy(patch_outputs).share_memory_()
+        self._mp_shared_space.patch_inputs = patch_inputs
+        self._mp_shared_space.patch_outputs = patch_outputs
+        self._mp_shared_space.wsi_idx = torch.Tensor([wsi_idx]).share_memory_()
+
+        # ! TODO: need a protocol for pbar, or a decorator to make this less redundant
+        pbar_desc = "Process Batch: "
+        pbar = tqdm.tqdm(
+            desc=pbar_desc,
+            leave=True,
+            total=int(len(loader)),
+            ncols=80,
+            ascii=True,
+            position=0,
+        )
+
+        cum_output = []
+        for _, batch_data in enumerate(loader):
+            sample_datas, sample_infos = batch_data
+            batch_size = sample_infos.shape[0]
+            # ! depending on the protocol of the output within infer_batch
+            # ! this may change, how to enforce/document/expose this in a
+            # ! sensible way?
+
+            # assume to return a list of L output,
```

Comment:
`L output` is a bit ambiguous - can we be more clear here. Also do we have a test to make sure the output format is correct?

Reply:
Likely into another PR once I have an idea to make a whole protocol for this. This relates to patch predictor also.